src/nupic/research/connections.py

```diff
@@ -251,50 +487,90 @@ def computeActivity(self, activeInput, activePermanenceThreshold,
+                                               iteration variable
 
     @return (tuple) Contains:
                       `activeSegments`         (list),
-                      `matchingSegments`        (list),
+                      `matchingSegments`       (list),
 
     Notes:
-      activeSegments and matchingSegments are sorted by the cell they are on.
+      activeSegments and matchingSegments are sorted by the cell they are on
+      and they are lists of SegmentOverlap objects.
     """
 
-    numActiveSynapsesForSegment = [0] * self._nextSegmentIdx
-    numMatchingSynapsesForSegment = [0] * self._nextSegmentIdx
+    numActiveSynapsesForSegment = [0] * self._nextFlatIdx
+    numMatchingSynapsesForSegment = [0] * self._nextFlatIdx
 
     for cell in activeInput:
-      for synapseData in self.synapsesForPresynapticCell(cell).values():
-        segment = synapseData.segment
+      for synapse in self.synapsesForPresynapticCell(cell):
+        synapseData = self.dataForSynapse(synapse)
+        segment = synapse.segment
         permanence = synapseData.permanence
-        if permanence >= matchingPermananceThreshold:
-          numMatchingSynapsesForSegment[segment] += 1
-          if synapseData.permanence >= activePermanenceThreshold:
-            numActiveSynapsesForSegment[segment] += 1
+        segmentData = self.dataForSegment(segment)
+        if permanence - matchingPermananceThreshold > -EPSILON:
+          numMatchingSynapsesForSegment[segmentData.flatIdx] += 1
+          if permanence - activePermanenceThreshold > -EPSILON:
+            numActiveSynapsesForSegment[segmentData.flatIdx] += 1
+
+    if recordIteration:
+      self._iteration += 1
 
     activeSegments = []
     matchingSegments = []
-    for i in xrange(self._nextSegmentIdx):
-      if numActiveSynapsesForSegment[i] >= activeSynapseThreshold:
-        activeSegments.append(i)
+    for i in xrange(self._nextFlatIdx):
+      segment = self._segmentForFlatIdx[i]
+      numActive = numActiveSynapsesForSegment[i]
+      if numActive >= activeSynapseThreshold:
+        segmentOverlap = SegmentOverlap(segment, numActive)
+        activeSegments.append(segmentOverlap)
+        
+        if recordIteration:
+          self.dataForSegment(segment).lastUsedIteration = self._iteration
+
 
-    for i in xrange(self._nextSegmentIdx):
-      if numMatchingSynapsesForSegment[i] >= matchingSynapseThreshold:
-        matchingSegments.append(i)
+    for i in xrange(self._nextFlatIdx):
+      segment = self._segmentForFlatIdx[i]
+      numMatching = numMatchingSynapsesForSegment[i]
+      if numMatching >= matchingSynapseThreshold:
+        segmentOverlap = SegmentOverlap(segment, numMatching)
+        matchingSegments.append(segmentOverlap)
 
-    segCmp = lambda a, b: self._segments[a] - self._segments[b]
-    return (sorted(activeSegments, cmp = segCmp),
-            sorted(matchingSegments, cmp = segCmp))
+    segCmp = lambda s: s.segment
```

Comment:
`segCmp` is no longer the right name. I think it'd be fine to just do inline lambdas in the `sorted` call.